### PR TITLE
Added test for unsafe functions not evaluating in --audit mode

### PR DESCRIPTION
**Pull request by craigcomstock**
Ticket: ENT-5299
Changelog: none

### CODE REVIEW

**Comment by craigcomstock** [comment]
Have left this as-is without the `--audit` option since that's not here yet. Wanted to get some feedback before squashing.

**Comment by craigcomstock** [comment]
squashed commits. @nickanderson can you take a look?

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf [hunk 1]
```diff
@@ -0,0 +1,16 @@
+bundle agent __main__
```
Comment (nickanderson) [inline tests/acceptance/29_audit_mode/audit_safe_functions.cf:1]: I believe this should include dcs.sub?

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  vars:
+    "cf_agent" string => ifelse(isvariable("sys.cf_agent"), "$(sys.cf_agent)", "/var/cfengine/bin/cf-agent");
+#    "result" string => execresult("$(cf_agent) --audit -Kf $(this.promise_filename).sub", useshell);
+    "result" string => execresult("$(cf_agent) -Kf $(this.promise_filename).sub", useshell);
+
+  reports:
+    "result is $(result)";
+    "$(this.promise_filename) Pass"
+      if => regcmp("R:.* Pass", "$(result)");
+    "$(this.promise_filename) FAIL"
```
Comment (nickanderson) [inline tests/acceptance/29_audit_mode/audit_safe_functions.cf:12]: I think that you can use one of the dcs bundles for this.

Perhaps

- bundle agent =dcs_check_regcmp(regex, thestring, test, expected_mismatch)= :: Pass/FAIL
     based on =expected_mismatch= (=yes=/=no=) when the regular expression
     =regex= matches the string =thestring=.

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf.sub [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+{
+  classes:
+    "pass_$(test.unsafe_functions)" expression => not(fileexists("$(sys.statedir)$(const.dirsep)$(test.unsafe_functions)_unsafe"));
+    "fail_$(test.unsafe_functions)" expression => fileexists("$(sys.statedir)$(const.dirsep)$(test.unsafe_functions)_unsafe");
+
+  reports:
+    all_pass::
+      "all_pass is defined!";
+    any::
+    "$(this.promise_filename) SFAIL"
```
Comment (nickanderson) [inline tests/acceptance/29_audit_mode/audit_safe_functions.cf.sub:71]: When I have done tests for sub executions, the sub policy files I tend to make very simple, not returning Sfail or whatever, i do all that from the parent. Maybe I  am missing something hree.

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf.sub [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      perms => m(700),
+      content => "#!/bin/bash
+touch $(sys.statedir)/usemodule_safe
+";
+}
+
+bundle agent test
+{
+  meta:
+    "test_soft_fail"
```
Comment (nickanderson) [inline tests/acceptance/29_audit_mode/audit_safe_functions.cf.sub:32]: I think this meta data should be up in the parent.

**Review by nickanderson — CHANGES_REQUESTED** [review]

**Review by nickanderson — DISMISSED** [review]

**Comment by nickanderson** [comment]
All of my comments were more about the test framework, not about the actual testing. That part is ok. If you think it's better to bury the details and metadata in the subtest and not use DCS, ok. I will opena  pr to remove the instruction about not using dcs.cf.sub

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf [hunk 5]
```diff
@@ -0,0 +1,16 @@
+bundle agent __main__
+{
+  vars:
+    "cf_agent" string => ifelse(isvariable("sys.cf_agent"), "$(sys.cf_agent)", "/var/cfengine/bin/cf-agent");
+#    "result" string => execresult("$(cf_agent) --audit -Kf $(this.promise_filename).sub", useshell);
```
Comment (vpodzime) [inline tests/acceptance/29_audit_mode/audit_safe_functions.cf:5]: It will be `--audit=manifest` or `--audit=diff`.

**Review by vpodzime — COMMENTED** [review]
Looks good to me, generally. Let's merge this together with the `--audit=manifest` implementation so that we can actually try running the test.

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf (hunk 5 above)
Comment (craigcomstock) [reply tests/acceptance/29_audit_mode/audit_safe_functions.cf:5]: ack

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf [hunk 6]
```diff
@@ -0,0 +1,16 @@
+bundle agent __main__
+{
+  vars:
+    "cf_agent" string => ifelse(isvariable("sys.cf_agent"), "$(sys.cf_agent)", "/var/cfengine/bin/cf-agent");
```
[the remark this answers is not in the document]
Comment (craigcomstock) [reply tests/acceptance/29_audit_mode/audit_safe_functions.cf:4]: It does seem that when I run cf-agent in-source-tree (the libtool script really) then `sys.cf_agent` is not defined. I believe I started using this when locally testing `tests/acceptance/28_inform_testing`.

**Review by nickanderson — DISMISSED** [review]
LGTM

File: tests/acceptance/29_audit_mode/audit_safe_functions.cf [hunk 7]
```diff
@@ -0,0 +1,21 @@
+body common control
+{
+  inputs => { "../default.cf.sub" };
+  bundlesequence => { default("$(this.promise_filename)") };
+}
+
+
+bundle agent test
+{
+  meta:
```
Comment (nickanderson) [inline tests/acceptance/29_audit_mode/audit_safe_functions.cf:10]: Oh, can you add

```
"description" -> { "ENT-5299" }
  string => "Test that unsafe function sare only evaulated in --audit mode when tagged as audit_safe";
```

Comment (craigcomstock) [reply tests/acceptance/29_audit_mode/audit_safe_functions.cf:10]: yup

Comment (craigcomstock) [reply tests/acceptance/29_audit_mode/audit_safe_functions.cf:10]: like line 14 @nickanderson ?

Comment (nickanderson) [reply tests/acceptance/29_audit_mode/audit_safe_functions.cf:10]: yep!

**Review by nickanderson — DISMISSED** [review]

**Review by nickanderson — APPROVED** [review]

**Comment by nickanderson** [comment]
@vpodzime it's soft failing now, so it should cause your manifest pr to fail when it starts passing (or unset soft fail with your pr)

**Pull request merged**